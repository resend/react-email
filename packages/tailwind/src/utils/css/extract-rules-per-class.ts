import { type CssNode, type Rule, string, walk } from 'css-tree';
import { isRuleInlinable } from './is-rule-inlinable';

export function extractRulesPerClass(root: CssNode, classes: string[]) {
  const classSet = new Set(classes);

  const inlinableRules = new Map<string, Rule>();
  const nonInlinableRules = new Map<string, Rule>();
  walk(root, {
    visit: 'Rule',
    enter(rule) {
      const selectorClasses: string[] = [];
      walk(rule, {
        visit: 'ClassSelector',
        enter(classSelector) {
          selectorClasses.push(string.decode(classSelector.name));
        },
      });
      if (isRuleInlinable(rule)) {
        for (const className of selectorClasses) {
          if (classSet.has(className)) {
            inlinableRules.set(className, rule);
          }
        }
      } else {
        for (const className of selectorClasses) {
          if (classSet.has(className)) {
            nonInlinableRules.set(className, rule);
          }
        }
      }
    },
  });
  return {
    inlinable: inlinableRules,
    nonInlinable: nonInlinableRules,
  };
}
