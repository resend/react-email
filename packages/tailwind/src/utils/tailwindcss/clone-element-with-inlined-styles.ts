import type { Node } from 'postcss';
import React from 'react';
import type { EmailElementProps } from '../../tailwind';
import { sanitizeClassName } from '../compatibility/sanitize-class-name';
import { makeInlineStylesFor } from '../css/make-inline-styles-for';
import { sanitizeMediaQueries } from '../css/media-queries/sanitize-media-queries';
import { sanitizeDeclarations } from '../css/sanitize-declarations';
import { isComponent } from '../react/is-component';
import type { setupTailwind } from './setup-tailwind';

export const cloneElementWithInlinedStyles = (
  element: React.ReactElement<EmailElementProps>,
  tailwind: ReturnType<typeof setupTailwind>,
) => {
  const propsToOverwrite: Partial<EmailElementProps> = {};

  let nonInlinableClasses: string[] = [];
  let nonInlineStyleNodes: Node[] = [];

  if (element.props.className) {
    const rootForClasses = tailwind.generateRootForClasses(
      element.props.className.split(' '),
    );
    sanitizeDeclarations(rootForClasses);

    const { sanitizedAtRules, mediaQueryClasses } =
      sanitizeMediaQueries(rootForClasses);
    nonInlinableClasses = mediaQueryClasses;
    nonInlineStyleNodes = sanitizedAtRules;

    const { styles, residualClassName } = makeInlineStylesFor(
      element.props.className,
      rootForClasses,
    );
    propsToOverwrite.style = {
      ...styles,
      ...element.props.style,
    };

    if (!isComponent(element)) {
      if (residualClassName.trim().length > 0) {
        propsToOverwrite.className = residualClassName;

        /*
          We sanitize only the class names of Tailwind classes that we are not going to inline
          to avoid unpredictable behavior on the user's code. If we did sanitize all classes
          a user-defined class could end up also being sanitized which would lead to unexpected 
          behavior and bugs that are hard to track.
        */
        for (const singleClass of mediaQueryClasses) {
          propsToOverwrite.className = propsToOverwrite.className.replace(
            singleClass,
            sanitizeClassName(singleClass),
          );
        }
      } else {
        propsToOverwrite.className = undefined;
      }
    }
  }

  const newProps = {
    ...element.props,
    ...propsToOverwrite,
  };

  return {
    elementWithInlinedStyles: React.cloneElement(
      element,
      newProps,
      newProps.children,
    ),

    nonInlinableClasses,
    nonInlineStyleNodes,
  };
};
