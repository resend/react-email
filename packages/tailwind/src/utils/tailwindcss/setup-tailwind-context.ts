import { createContext } from 'tailwindcss/lib/lib/setupContextUtils';
import resolveConfig from 'tailwindcss/resolveConfig';
import type { TailwindConfig } from '../../tailwind';

export const setupTailwindContext = (config: TailwindConfig) => {
  return createContext(
    resolveConfig({
      ...config,
      content: [],
      corePlugins: {
        preflight: false,
      },
    }),
  );
};
