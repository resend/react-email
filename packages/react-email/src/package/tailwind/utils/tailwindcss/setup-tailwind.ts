import type { Rule } from 'patched-postcss';
import parse from 'patched-postcss/lib/parse';
import collapseAdjacentRules from 'patched-tailwindcss/lib/lib/collapseAdjacentRules';
import collapseDuplicateDeclarations from 'patched-tailwindcss/lib/lib/collapseDuplicateDeclarations';
import evaluateTailwindFunctions from 'patched-tailwindcss/lib/lib/evaluateTailwindFunctions';
import expandApplyAtRules from 'patched-tailwindcss/lib/lib/expandApplyAtRules';
import expandTailwindAtRules from 'patched-tailwindcss/lib/lib/expandTailwindAtRules';
import { generateRules as rawGenerateRules } from 'patched-tailwindcss/lib/lib/generateRules';
import partitionApplyAtRules from 'patched-tailwindcss/lib/lib/partitionApplyAtRules';
import resolveDefaultsAtRules from 'patched-tailwindcss/lib/lib/resolveDefaultsAtRules';
import substituteScreenAtRules from 'patched-tailwindcss/lib/lib/substituteScreenAtRules';
import type { TailwindConfig } from '../../tailwind';
import { resolveAllCSSVariables } from '../css/resolve-all-css-variables';
import { setupTailwindContext } from './setup-tailwind-context';

const tailwindAtRulesRoot = parse(
  `
  @tailwind base;
  @tailwind components;
`,
).root();

export function setupTailwind(config: TailwindConfig) {
  const tailwindContext = setupTailwindContext(config);
  return {
    generateRootForClasses: (classes: string[]) => {
      const bigIntRuleTuples: [bigint, Rule][] = rawGenerateRules(
        new Set(classes),
        tailwindContext,
      );

      const root = tailwindAtRulesRoot
        .clone()
        .append(...bigIntRuleTuples.map(([, rule]) => rule));
      partitionApplyAtRules()(root);
      // This is fine because the internal await is never actually called out
      // because of there not being any `changedContent` files on the context
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      expandTailwindAtRules(tailwindContext)(root);
      partitionApplyAtRules()(root);
      expandApplyAtRules(tailwindContext)(root);
      evaluateTailwindFunctions(tailwindContext)(root);
      substituteScreenAtRules(tailwindContext)(root);
      resolveDefaultsAtRules(tailwindContext)(root);
      collapseAdjacentRules(tailwindContext)(root);
      collapseDuplicateDeclarations(tailwindContext)(root);

      resolveAllCSSVariables(root);

      return root;
    },
  };
}
