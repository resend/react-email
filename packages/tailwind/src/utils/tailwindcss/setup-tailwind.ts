import type { Rule } from 'postcss';
import { parse } from 'postcss';
import collapseAdjacentRules from 'tailwindcss/lib/lib/collapseAdjacentRules';
import collapseDuplicateDeclarations from 'tailwindcss/lib/lib/collapseDuplicateDeclarations';
import evaluateTailwindFunctions from 'tailwindcss/lib/lib/evaluateTailwindFunctions';
import expandApplyAtRules from 'tailwindcss/lib/lib/expandApplyAtRules';
import expandTailwindAtRules from 'tailwindcss/lib/lib/expandTailwindAtRules';
import { generateRules as rawGenerateRules } from 'tailwindcss/lib/lib/generateRules';
import partitionApplyAtRules from 'tailwindcss/lib/lib/partitionApplyAtRules';
import resolveDefaultsAtRules from 'tailwindcss/lib/lib/resolveDefaultsAtRules';
import substituteScreenAtRules from 'tailwindcss/lib/lib/substituteScreenAtRules';
import type { TailwindConfig } from '../../tailwind';
import { resolveAllCSSVariables } from '../css/resolve-all-css-variables';
import { setupTailwindContext } from './setup-tailwind-context';

const tailwindAtRulesRoot = parse(
  `
  @tailwind base;
  @tailwind components;
`,
).root();

export function setupTailwind(config: TailwindConfig) {
  const tailwindContext = setupTailwindContext(config);
  return {
    generateRootForClasses: (classes: string[]) => {
      const bigIntRuleTuples: [bigint, Rule][] = rawGenerateRules(
        new Set(classes),
        tailwindContext,
      );

      const root = tailwindAtRulesRoot
        .clone()
        .append(...bigIntRuleTuples.map(([, rule]) => rule));
      partitionApplyAtRules()(root);
      // This is fine because the internal await is never actually called out
      // because of there not being any `changedContent` files on the context
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      expandTailwindAtRules(tailwindContext)(root);
      partitionApplyAtRules()(root);
      expandApplyAtRules(tailwindContext)(root);
      evaluateTailwindFunctions(tailwindContext)(root);
      substituteScreenAtRules(tailwindContext)(root);
      resolveDefaultsAtRules(tailwindContext)(root);
      collapseAdjacentRules(tailwindContext)(root);
      collapseDuplicateDeclarations(tailwindContext)(root);

      resolveAllCSSVariables(root);

      return root;
    },
  };
}
