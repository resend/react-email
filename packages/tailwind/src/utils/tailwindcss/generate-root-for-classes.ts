import { parse } from 'css-tree';
import { type Config, compile } from 'tailwindcss';
import { resolveAllCSSVariables } from '../css/resolve-all-css-variables';
import indexCss from './tailwind-stylesheets/index';
import preflightCss from './tailwind-stylesheets/preflight';
import themeCss from './tailwind-stylesheets/theme';
import utilitiesCss from './tailwind-stylesheets/utilities';

const baseCss = `
@layer theme, base, components, utilities;
@import "tailwindcss/theme.css" layer(theme);
@import "tailwindcss/utilities.css" layer(utilities);
@config;
`;

export async function generateRootForClasses(
  classes: string[],
  config: Config,
) {
  const compiler = await compile(baseCss, {
    async loadModule(id, base, resourceHint) {
      if (resourceHint === 'config') {
        return {
          path: id,
          base: base,
          module: config,
        };
      }

      throw new Error(
        `NO-OP: should we implement support for ${resourceHint}?`,
      );
    },
    async loadStylesheet(id, base) {
      if (id === 'tailwindcss') {
        return {
          base,
          path: 'tailwindcss/index.css',
          content: indexCss,
        };
      }

      if (id === 'tailwindcss/preflight.css') {
        return {
          base,
          path: id,
          content: preflightCss,
        };
      }

      if (id === 'tailwindcss/theme.css') {
        return {
          base,
          path: id,
          content: themeCss,
        };
      }

      if (id === 'tailwindcss/utilities.css') {
        return {
          base,
          path: id,
          content: utilitiesCss,
        };
      }

      throw new Error(
        'stylesheet not supported, you can only import tailwindcss',
      );
    },
  });
  const css = compiler.build(classes);
  const root = parse(css);
  resolveAllCSSVariables(root);

  return root;
}
