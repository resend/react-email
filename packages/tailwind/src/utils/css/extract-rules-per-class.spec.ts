import { generate, type Rule } from 'css-tree';
import { setupTailwind } from '../tailwindcss/setup-tailwind';
import { extractRulesPerClass } from './extract-rules-per-class';

describe('extractRulesPerClass()', async () => {
  function convertToComparable(map: Map<string, Rule>): Record<string, string> {
    return Object.fromEntries(map.entries().map(([k, v]) => [k, generate(v)]));
  }

  it('works with just inlinable utilities', async () => {
    const tailwind = await setupTailwind({});
    const classes = ['text-center', 'bg-red-500'];
    tailwind.addUtilities(classes);

    const stylesheet = tailwind.getStyleSheet();

    const { inlinable, nonInlinable } = extractRulesPerClass(
      stylesheet,
      classes,
    );

    expect(convertToComparable(inlinable)).toMatchSnapshot();
    expect(convertToComparable(nonInlinable)).toMatchSnapshot();
  });

  it('should work with non-inlinable utilities', async () => {
    const tailwind = await setupTailwind({});
    const classes = ['lg:w-1/2'];
    tailwind.addUtilities(classes);

    const stylesheet = tailwind.getStyleSheet();

    const { inlinable, nonInlinable } = extractRulesPerClass(
      stylesheet,
      classes,
    );

    expect(convertToComparable(inlinable)).toMatchSnapshot();
    expect(convertToComparable(nonInlinable)).toMatchSnapshot();
  });

  it('should work with a mix of inlinable and non-inlinable utilities', async () => {
    const tailwind = await setupTailwind({});
    const classes = [
      'text-center',
      'bg-red-500',
      'some-other-class', // should be ignored
      'w-full',
      'lg:w-1/2',
    ];
    tailwind.addUtilities(classes);

    const stylesheet = tailwind.getStyleSheet();
    const { inlinable, nonInlinable } = extractRulesPerClass(
      stylesheet,
      classes,
    );
    expect(convertToComparable(inlinable)).toMatchSnapshot();
    expect(convertToComparable(nonInlinable)).toMatchSnapshot();
  });
});
