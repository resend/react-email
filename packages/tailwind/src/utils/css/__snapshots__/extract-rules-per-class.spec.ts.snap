// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractRulesPerClass() > handles a mix of inlinable and non-inlinable utilities 1`] = `
{
  "bg-red-500": ".bg-red-500{background-color:var(--color-red-500)}",
  "text-center": ".text-center{text-align:center}",
  "w-full": ".w-full{width:100%}",
}
`;

exports[`extractRulesPerClass() > handles a mix of inlinable and non-inlinable utilities 2`] = `
{
  "lg:w-1/2": ".lg\\:w-1\\/2{@media (width>=64rem){width:calc(1/2*100%)}}",
}
`;

exports[`extractRulesPerClass() > handles non-inlinable utilities 1`] = `{}`;

exports[`extractRulesPerClass() > handles non-inlinable utilities 2`] = `
{
  "lg:w-1/2": ".lg\\:w-1\\/2{@media (width>=64rem){width:calc(1/2*100%)}}",
}
`;

exports[`extractRulesPerClass() > works with just inlinable utilities 1`] = `
{
  "bg-red-500": ".bg-red-500{background-color:var(--color-red-500)}",
  "text-center": ".text-center{text-align:center}",
}
`;

exports[`extractRulesPerClass() > works with just inlinable utilities 2`] = `{}`;
